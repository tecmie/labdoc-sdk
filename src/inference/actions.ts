// import { trpc } from './trpc';

// @ts-ignore
// export const inference = trpc.scanner.read.useMutation();
