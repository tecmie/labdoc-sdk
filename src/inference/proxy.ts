import { createTRPCProxyClient, httpBatchLink, loggerLink } from '@trpc/client';
import { AnyRouter, ProcedureRouterRecord, AnyProcedure } from '@trpc/server';
import { DecorateProcedure } from '@trpc/react-query/shared';

type DecoratedProcedureRecord<
  TProcedures extends ProcedureRouterRecord,
  // eslint-disable-next-line @typescript-eslint/no-unused-vars
  TRouter extends AnyRouter
> = {
  [TKey in keyof TProcedures]: TProcedures[TKey] extends AnyRouter
    ? DecoratedProcedureRecord<
        TProcedures[TKey]['_def']['record'],
        TProcedures[TKey]
      >
    : TProcedures[TKey] extends AnyProcedure
    ? DecorateProcedure<TProcedures[TKey], any>
    : never;
};

export type CreateLabProxyOptions = {
  customHttpLink?: string;
  secretOrKey: string;
};

export const createLabProxy = ({
  secretOrKey,
  customHttpLink,
}: CreateLabProxyOptions): DecoratedProcedureRecord<any, AnyRouter> =>
  // @ts-ignore
  createTRPCProxyClient({
    links: [
      loggerLink({
        enabled: (opts) =>
          process.env.NODE_ENV === 'development' ||
          (opts.direction === 'down' && opts.result instanceof Error),
      }),
      httpBatchLink({
        // url: `http://localhost:3001/api/trpc`,
        url: customHttpLink ?? `https://labai.tecmie.africa/api/trpc`,
        /**
         * Headers will be called on each request.
         */
        headers: () => ({
          'x-secret': secretOrKey,
        }),
      }),
    ],
  });
