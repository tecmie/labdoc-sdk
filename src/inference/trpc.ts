import { createTRPCReact } from '@trpc/react-query';
import {
  type inferRouterInputs,
  type inferRouterOutputs,
  AnyRouter,
} from '@trpc/server';

type AppRouter = AnyRouter;
export const trpc = createTRPCReact<AppRouter>();

/**
 * Inference helper for inputs
 * @example type HelloInput = RouterInputs['example']['hello']
 * */
export type RouterInputs = inferRouterInputs<AppRouter>;
/**
 * Inference helper for outputs
 * @example type HelloOutput = RouterOutputs['example']['hello']
 * */
export type RouterOutputs = inferRouterOutputs<AppRouter>;
