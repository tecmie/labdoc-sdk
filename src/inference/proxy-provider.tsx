import * as React from 'react';
import { createContext, useContext } from 'react';
import { createLabProxy } from './proxy';

export const LabBaseContext = createContext(
  createLabProxy({ secretOrKey: '' })
);

type LabProviderProps = {
  children: React.ReactNode;
  customHttpLink?: string;
  secretOrKey: string;
};

export const LabProvider = ({
  secretOrKey,
  customHttpLink,
  children,
}: LabProviderProps) => {
  const proxy = createLabProxy({ secretOrKey, customHttpLink });

  return (
    <LabBaseContext.Provider value={proxy}>{children}</LabBaseContext.Provider>
  );
};

export const useIn = () => {
  const context = useContext(LabBaseContext);

  if (!context) {
    throw new Error('useScanner must be used within a LabProvider');
  }

  return context;
};
