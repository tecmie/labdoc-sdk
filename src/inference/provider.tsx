import { httpBatchLink, loggerLink } from '@trpc/client';
import { QueryClient, QueryClientProvider } from '@tanstack/react-query';
import React, { useState } from 'react';
import superjson from 'superjson';
import { trpc } from './trpc';

type ProviderProps = {
  children: React.ReactNode;
  defaultQueryClient?: QueryClient;
  httpRpcURL?: string;
  secret: string;
};

export const LabAIClientProvider = ({
  children,
  defaultQueryClient,
  httpRpcURL,
  secret,
}: ProviderProps) => {
  const getBaseUrl = () => {
    if (typeof window !== 'undefined') return ''; // browser should use relative url
    return httpRpcURL || `https://labai.tecmie.africa`; // dev SSR should use localhost
  };

  const secretOrKey = secret ?? process.env.INFERENCE_SECRET;
  if (!secretOrKey) {
    throw new Error('secret is required');
  }

  const [queryClient] = useState(() => defaultQueryClient ?? new QueryClient());

  const [trpcClient] = useState(() =>
    trpc.createClient({
      transformer: superjson,
      links: [
        loggerLink({
          enabled: (opts) =>
            process.env.NODE_ENV === 'development' ||
            (opts.direction === 'down' && opts.result instanceof Error),
        }),
        httpBatchLink({
          url: `${getBaseUrl()}/api/trpc`,
          /**
           * Headers will be called on each request.
           */
          headers: () => ({
            'x-secret': secretOrKey,
          }),
        }),
      ],
    })
  );
  return (
    <trpc.Provider client={trpcClient} queryClient={queryClient}>
      <QueryClientProvider client={queryClient}>{children}</QueryClientProvider>
    </trpc.Provider>
  );
};
