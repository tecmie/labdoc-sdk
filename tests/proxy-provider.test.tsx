import * as React from 'react';
import { render } from '@testing-library/react';
import { LabProvider } from '../src/inference/proxy-provider';

require('isomorphic-fetch'); // Adjust the path as necessary

test('LabProvider matches snapshot', () => {
  const { asFragment } = render(
    <LabProvider secretOrKey="test-key">
      <div>Test child</div>
    </LabProvider>
  );

  expect(asFragment()).toMatchSnapshot();
});
