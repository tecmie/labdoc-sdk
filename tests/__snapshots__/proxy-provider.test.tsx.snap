// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LabProvider matches snapshot 1`] = `
<DocumentFragment>
  <div>
    Test child
  </div>
</DocumentFragment>
`;
